### PR TITLE
Make HTTP response calling error callback on 400 or 500

### PR DESCRIPTION
**Pull request by aliok**
## Motivation

While working on https://issues.jboss.org/browse/AEROGEAR-2395 , I noticed the metrics related integration tests pass all the time.

As written in  <https://square.github.io/okhttp/3.x/okhttp/okhttp3/Call.html#execute--> , execute method does not throw an exception when status is 400 or 500. It only throws an exception when there's a connection problem.

But dependent code assumes that `onError` is called when 400 or 500.

To test the wrong behavior, just run the integration tests without the change in the PR. Even when the URL in integration-test-mobile-services.json is something that returns 500, it will pass. Just copy paste metrics service url, but change a letter so that it returns "503 - Service unavailable".

### CODE REVIEW

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java [hunk 1]
```diff
@@ -25,9 +25,19 @@
...
                 response = okHttpCall.execute();
                 requestCompleteLatch.countDown();
-                runSuccessHandler();
+
+                if(response.isSuccessful()){
+                    // status 200 or 300
+                    runSuccessHandler();
+                } else {
+                    // status 400 or 500
+                    error = new Exception(response.message());
```
Comment (aliok) [inline core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: @danielpassos any better ideas here instead of plain Exception?

**Comment by aliok** [comment]
@danielpassos @secondsun I have a question about redirects. Should the HTTP client follow redirects? It doesn't seem it does that. This is independent from my PR here.

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java (hunk 1 above)
Comment (danielpassos) [reply core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: What about raise a [HttpException](https://github.com/aerogear/aerogear-android-sdk/blob/master/core/src/main/java/org/aerogear/mobile/core/exception/HttpException.java) instead and update the HttpException to receive the Http status code on the constructor?

**Comment by danielpassos** [comment]
I think we can do the redirect in our library, we handle it for push on the old push library. Wdyt @secondsun?

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java (hunk 1 above)
Comment (aliok) [reply core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: ok, didn't notice we have that one. fixing it now.

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java [hunk 2]
```diff
@@ -90,8 +101,8 @@ public HttpResponse onError(Runnable errorHandler) {
     @Override
     public HttpResponse onSuccess(Runnable successHandler) {
         this.successHandler = successHandler;
-        // If there is _any_ response the success handler should run
-        if (response != null) {
+        // If there is _any_ response and it is successful the success handler should run immediately
+        if (response != null && (response.isSuccessful() || response.isRedirect())) {
             runSuccessHandler();
```
Comment (aliok) [inline core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: As the success handler is immediately called upon setting a success handler using `onSuccess`, I needed to adjust the condition to run it. In some cases, I saw the response is received faster than execution arrives at setting `onSuccess` handler.

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java [hunk 3]
```diff
@@ -108,9 +119,12 @@ public void waitForCompletionAndClose() {
...
             // If a success Handler was set before this wait then we need to make
             // sure that gets called before we free these resources.
-            if (error == null) {
+            if (error == null && (response.isSuccessful() || response.isRedirect())) {
                 runSuccessHandler();
             } else {
+                if (error == null) {
+                    this.error = new HttpException(response.code(), response.message());
+                }
                 runErrorHandler();
```
Comment (aliok) [inline core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: Another case where things fail timing happens differently.

**Comment by aliok** [comment]
@danielpassos this PR is ready for the final review and the merge.

File: core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java [hunk 4]
```diff
@@ -90,8 +101,9 @@ public HttpResponse onError(Runnable errorHandler) {
     @Override
     public HttpResponse onSuccess(Runnable successHandler) {
         this.successHandler = successHandler;
-        // If there is _any_ response the success handler should run
-        if (response != null) {
+        // If there is _any_ response and it is successful the success handler should run
+        // immediately
+        if (response != null && (response.isSuccessful() || response.isRedirect())) {
             runSuccessHandler();
```
Comment (aliok) [inline core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java]: As the success handler is immediately called upon setting a success handler using `onSuccess`, I needed to adjust the condition to run it. In some cases, I saw the response is received faster than execution arrives at setting `onSuccess` handler.

**Comment by danielpassos** [comment]
@aliok I'm ok with ignore the Jenkins for now and move on if you want.

**Comment by aliok** [comment]
@danielpassos ok, let's merge this one. unit tests pass, integration tests fail.

**Pull request merged**